File: 22.generate-parentheses.go
// @leet start
func recursiveGenerateParenthesis(left, right int, path string, res *[]string) {
	if left == 0 && right == 0 {
		*res = append(*res, path)
	}
	if left > 0 {
		recursiveGenerateParenthesis(left-1, right, path+"(", res)
	}
	if right > left {
		recursiveGenerateParenthesis(left, right-1, path+")", res)
	}
}

func generateParenthesis(n int) []string {
	res := []string{}
	recursiveGenerateParenthesis(n, n, "", &res)
	return res
}

// @leet end

File: 22.generate-parentheses.js
// @leet start
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
  let res = [];
  let backtrack = (open, close, s) => {
    if (open === n && close === n) {
      res.push(s);
    }
    if (open < n) {
      backtrack(open + 1, close, s + "(");
    }
    if (close < open) {
      backtrack(open, close + 1, s + ")");
    }
  }
  backtrack(0, 0, "");
  return res;
};
// @leet end
